ytt/001-schemas/README.md

```diff
@@ -239,111 +900,393 @@ Beyond specifying a type for a value, one can specify more dynamic constraints o
   - base64.decodable()
   - json.decodable()
   - yaml.decodable()
+  - regexp.is_valid()
 
   String values, by default, have `@schema/validate min_len=1` validation added.
 
-### Describing annotations
+## Part 8: Interactions with Existing Features
 
-#### `@schema/title`
+### Changes to Data Values
 
-  This annotation provides a way to add a short title to the section of a schema associated with a key, similar to the [JSON schema title field](https://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.9.1).
+Prior to Schemas, Data Values protected against mistakes like typo'ed key names through the
+strict merge of additional data values documents: to successfully merge a new key, it must be marked
+as `@overlay/match missing_ok=True`.
 
-  ```yaml
-  #@schema/title "User Password"
-  user_password: ""
-  ```
+The responsibility to ensure key names are valid are shifted from Data Values to Schema.
 
-  If the annotation is not present, the title will be inferred from the key name by replacing special characters with spaces and capitalizing the first letter similar to [rails humanize functionality](https://apidock.com/rails/String/humanize).
+When schema is present, Data Values documents imply an automatic merge of new keys. This is 
+equivalent to all `@data/values` documents being annotated with `@overlay/match-child-defaults missing_ok=True`.
 
-#### `@schema/doc`
+### Overlays of Schema
 
-  This annotation is a way to add a longer description to the section of a schema associated with a key, similar to the JSON Schema description field.
+TBD
 
-  ```yaml
-  #@schema/doc "The user password used to log in to the system"
-  user_password: ""
-  ```
+- It must be possible to extend Schema (i.e. add new items)
+- Is it necessary to be able to modify Schema?
 
-#### `@schema/example`
 
-  This annotation will take one argument that is an example of a value that satisfies the schema.
+## Open Questions
 
-  ```yaml
-  #@schema/example "my_domain.example.com"
-  system_domain: ""
-  ```
+- It is not uncommon for a validation to want to refer to a sibling node.
+    - right now, the Configuration Author places these on the parent node.
+    - that puts separation between the node being constrained and its constraint.
+    - is there a feasible way for a validation rule to get access to its siblings?
```

Comment:
Answer: not now; we see no current way to accomplish this without undesirable consequences.

- in general, a node in a tree with knowledge of anything except its children leads to undue [connascence](https://en.wikipedia.org/wiki/Connascence) and [action at a distance](https://en.wikipedia.org/wiki/Action_at_a_distance_(computer_programming)).
- the physical distance between the target node and its parent is usually small enough that the resulting design is still readable/maintainable.
- the underlying need is greater than this simple example. It's common to want to express validation rules across whole branches (think dotted lines between tree nodes, making the dependency structure a graph).
  - the first order solution is to push the logic up the tree to the lowest common parent; this approach minimizes coupling since the parent _already_ has access to itself children. This is the current recommended solution.
  - when that approach introduces too much distance between two components that have a natural affinity (i.e. the lowest common parent is too far from the children or there are too many such needs), the current best known approach to this problem is to invert the dependencies so that they may be injected rather than referenced.
    - we do not yet understand enough about how this pattern will unfold in configuration management in `ytt` to pull the trigger on a dependency injection solution.
    - ideally this sort of mechanism follows or coincides with a component model that necessarily includes articulating [Service Provider Interfaces](https://en.wikipedia.org/wiki/Service_provider_interface) (SPI) and a mechanism for mating service needs with service providers (prior art example: [OSGi](https://en.wikipedia.org/wiki/OSGi)).